packages/react-reconciler/src/ReactFiberCacheComponent.new.js

```diff
@@ -57,6 +62,71 @@ let pooledCache: Cache | null = null;
 // cache from the render that suspended.
 const prevFreshCacheOnStack: StackCursor<Cache | null> = createCursor(null);
 
+// Creates a new empty Cache instance with a ref-count of 0. The caller is responsible
+// for retaining the cache once it is in use (retainCache), and releasing the cache
+// once it is no longer needed (releaseCache).
+let _cacheIndex = 0;
+export function createCache(): Cache {
+  const cache: Cache = {
+    controller: new AbortController(),
+    data: new Map(),
+    refCount: 0,
+  };
+
+  // TODO: remove debugging code
+  if (__DEV__) {
+    cache.key = '' + _cacheIndex++;
+    trace(`createCache #${cache.key ?? ''}`);
+  }
+
+  return cache;
+}
+
+export function retainCache(cache: Cache) {
+  if (__DEV__) {
+    trace(
+      `retainCache #${cache.key ?? ''} ${cache.refCount} -> ${cache.refCount +
+        1}`,
+    );
+  }
+  cache.refCount++;
+}
+
+// Cleanup a cache instance, potentially freeing it if there are no more references
+export function releaseCache(cache: Cache) {
+  if (__DEV__) {
+    trace(
+      `releaseCache #${cache.key ?? ''} ${cache.refCount} -> ${cache.refCount -
+        1}`,
+    );
+  }
+  cache.refCount--;
+  if (__DEV__) {
+    if (cache.refCount < 0) {
+      throw new Error(
+        'Error in React: cache reference count should not be negative',
+      );
+    }
+  }
+  if (cache.refCount === 0) {
+    // TODO: considering scheduling this call, and adding error handling for
+    // any event listeners that get triggered.
+    cache.controller.abort();
```

Comment:
What's the best way to schedule this? In discussion @acdlite mentioned that we don't want people to rely on the cleanup running too soon, but i also wonder if we just don't want people depending on _any_ specific timing. Thoughts on randomizing the timing? eg sometimes it occurs during passive effects, sometimes after? That gives us the most freedom to change the timing later wo breaking assumptions.

Reply:
I was thinking we'd use a Scheduler task. In practice I think it would always fire after passive effects because that's where the `release` happens. (Unless there are no passive effects and the final release is cache pool.) But it's not completely predictable because other events could squeeze in before it fires, including user inputs or paint.

One question is what priority to give the task. You don't want to wait too long if there are open connections that could be closed, or (in the future) if there are subscriptions that could be disposed. So I think it should be higher priority than Idle, which we use for offscreen rendering. What I'm less sure about is whether it should be higher priority than transitions. My guess is it doesn't need to be since transitions are not usually CPU bound for longer than it would matter for clean-up purposes; they should either finish or suspend and become I/O bound, which is an ideal time to do clean-up stuff.

Reply:
That overall makes sense. Thoughts on whether to randomize in dev, though? We could randomly switch between running cleanup during passive effects, scheduling at just above idle priority, or at idle priority.